Applications/Spire/Include/Spire/Ui/AssociativeValueModel.hpp

```diff
@@ -0,0 +1,242 @@
+       * @return Acceptable if the value was set successfully, Invalid
+       *         otherwise.
+       */
+      QValidator::State set_current(const Type& value) override;
+
+      /** Connects a slot to the current signal. */
+      boost::signals2::connection connect_current_signal(
+        const typename CurrentSignal::slot_type& slot) const override;
+
+    private:
+      mutable CurrentSignal m_current_signal;
+      Type m_current;
+      Type m_default_value;
+      std::unordered_map<Type, std::shared_ptr<BooleanModel>> m_models;
+      std::unordered_map<Type, boost::signals2::scoped_connection>
+        m_connections;
+      std::queue<Type> m_current_queue;
+      bool m_is_blocked;
+
+      void set_associated_model_value(const Type& value, bool model_value);
+      void on_current(const Type& value, bool is_selected);
+  };
+
+  template<typename T>
+  AssociativeValueModel<T>::AssociativeValueModel()
+    : AssociativeValueModel(Type{}) {}
+
+  template<typename T>
+  AssociativeValueModel<T>::AssociativeValueModel(Type default_value)
+    : AssociativeValueModel(std::make_shared<LocalBooleanModel>(true),
+        default_value) {}
+
+  template<typename T>
+  AssociativeValueModel<T>::AssociativeValueModel(
+      std::shared_ptr<BooleanModel> model, Type default_value)
+      : m_default_value(std::move(default_value)),
+        m_is_blocked(false) {
+    associate(model, m_default_value);
+  }
+
+  template<typename T>
+  void AssociativeValueModel<T>::associate(
+      const std::shared_ptr<BooleanModel>& model, const Type& value) {
+    m_models.insert_or_assign(value, model);
+    model->set_current(value == m_current);
+    m_connections.insert_or_assign(value, model->connect_current_signal(
+      [=] (auto is_selected) { on_current(value, is_selected); }));
+  }
+
+  template<typename T>
+  std::shared_ptr<BooleanModel> AssociativeValueModel<T>::make_association(
+      const Type& value) {
+    auto model = std::make_shared<LocalBooleanModel>(false);
+    associate(model, value);
+    return model;
+  }
+
+  template<typename T>
+  std::shared_ptr<BooleanModel> AssociativeValueModel<T>::find(
+      const Type& value) const {
+    auto i = m_models.find(value);
+    if(i == m_models.end()) {
+      return nullptr;
+    }
+    return i->second;
+  }
+
+  template<typename T>
+  QValidator::State AssociativeValueModel<T>::get_state() const {
+    return QValidator::Acceptable;
+  }
+
+  template<typename T>
+  const typename AssociativeValueModel<T>::Type&
+      AssociativeValueModel<T>::get_current() const {
+    return m_current;
+  }
+
+  template<typename T>
+  QValidator::State AssociativeValueModel<T>::set_current(const Type& value) {
+    if(m_models.find(value) == m_models.end()) {
+      return QValidator::Invalid;
+    }
+    if(m_is_blocked) {
+      m_current_queue.push(value);
+      return QValidator::Intermediate;
+    }
+    m_is_blocked = true;
+    auto previous = m_current;
```

Comment:
Feel free to remove this check ```previous != m_current```.

Reply:
Wouldn't this be the equivalent of always turning a light off and then back on to make sure it's on? The model ensures that the model associated with the current is always true, so there should never be a case when set_current is called with the existing current value and an associated value model would have to be updated.

Reply:
This is a good point, the issue is that this method is not re-entrant and so exposes a lot of inconsistencies. If you can come up with a solution that would be great, if not then I'll have to work on it or I will assign it out, the issue is I have a lot of tasks I need to finish up before I can work on anything new.